features/include/pcl/features/impl/gasd.hpp

```diff
@@ -0,0 +1,414 @@
+        const float v_xyzh1101 = v_xyz110 * coords[3];
+        const float v_xyzh1100 = v_xyz110 - v_xyzh1101;
+        const float v_xyzh1011 = v_xyz101 * coords[3];
+        const float v_xyzh1010 = v_xyz101 - v_xyzh1011;
+        const float v_xyzh1001 = v_xyz100 * coords[3];
+        const float v_xyzh1000 = v_xyz100 - v_xyzh1001;
+        const float v_xyzh0111 = v_xyz011 * coords[3];
+        const float v_xyzh0110 = v_xyz011 - v_xyzh0111;
+        const float v_xyzh0101 = v_xyz010 * coords[3];
+        const float v_xyzh0100 = v_xyz010 - v_xyzh0101;
+        const float v_xyzh0011 = v_xyz001 * coords[3];
+        const float v_xyzh0010 = v_xyz001 - v_xyzh0011;
+        const float v_xyzh0001 = v_xyz000 * coords[3];
+        const float v_xyzh0000 = v_xyz000 - v_xyzh0001;
+
+        hists[grid_idx][h_idx] += v_xyzh0000;
+        hists[grid_idx][h_idx + 1] += v_xyzh0001;
+        hists[grid_idx + 1][h_idx] += v_xyzh0010;
+        hists[grid_idx + 1][h_idx + 1] += v_xyzh0011;
+        hists[grid_idx + (grid_size + 2)][h_idx] += v_xyzh0100;
+        hists[grid_idx + (grid_size + 2)][h_idx + 1] += v_xyzh0101;
+        hists[grid_idx + (grid_size + 3)][h_idx] += v_xyzh0110;
+        hists[grid_idx + (grid_size + 3)][h_idx + 1] += v_xyzh0111;
+        hists[grid_idx + (grid_size + 2) * (grid_size + 2)][h_idx] += v_xyzh1000;
+        hists[grid_idx + (grid_size + 2) * (grid_size + 2)][h_idx + 1] += v_xyzh1001;
+        hists[grid_idx + (grid_size + 2) * (grid_size + 2) + 1][h_idx] += v_xyzh1010;
+        hists[grid_idx + (grid_size + 2) * (grid_size + 2) + 1][h_idx + 1] += v_xyzh1011;
+        hists[grid_idx + (grid_size + 3) * (grid_size + 2)][h_idx] += v_xyzh1100;
+        hists[grid_idx + (grid_size + 3) * (grid_size + 2)][h_idx + 1] += v_xyzh1101;
+        hists[grid_idx + (grid_size + 3) * (grid_size + 2) + 1][h_idx] += v_xyzh1110;
+        hists[grid_idx + (grid_size + 3) * (grid_size + 2) + 1][h_idx + 1] += v_xyzh1111;
+      }
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDEstimation<PointInT, PointOutT>::copyShapeHistogramsToOutput (const size_t grid_size,
+                                                                    const size_t hists_size,
+                                                                    const std::vector<Eigen::VectorXf> &hists,
+                                                                    PointCloudOut &output,
+                                                                    size_t &pos)
+  {
+    for (size_t i = 0; i < grid_size; ++i)
+    {
+      for (size_t j = 0; j < grid_size; ++j)
+      {
+        for (size_t k = 0; k < grid_size; ++k)
+        {
+          const size_t idx = ( (i + 1) * (grid_size + 2) + (j + 1)) * (grid_size + 2) + (k + 1);
+
+          std::copy (hists[idx].data () + 1, hists[idx].data () + hists_size + 1, output.points[0].histogram + pos);
```

Comment:
the read iterators look weird, especially the end one. Can you confirm that this is yielding the correct result?